rust/libceed/src/operator.rs

```diff
@@ -20,18 +20,226 @@
 
 use crate::prelude::*;
 
+// -----------------------------------------------------------------------------
+// CeedOperator Field context wrapper
+// -----------------------------------------------------------------------------
+#[derive(Debug)]
+pub struct OperatorField<'a> {
+    ceed: &'a crate::Ceed,
+    pub(crate) ptr: bind_ceed::CeedOperatorField,
+    elem_restriction: crate::ElemRestriction<'a>,
+    basis: crate::Basis<'a>,
+    vector: crate::Vector<'a>,
```

Comment:
Is it necessary that all these fields get mirrored into the Rust object rather than referenced on-demand through the API?

Compare `QFunctionField`, which only contains the pointer and thus can readily be used in slices a la f3e3b2465013c0e26ea7151c50f20d31a15b3659

Reply:
When I tried something similar with the operator fields, I was able to declare the returned slice as mut and start changing the underlying objects on the C side, which is why switched to only returning *Opts from the field, but an *Opt is only a ref and the actual Rest object needs to be owned somewhere

Reply:
Do you have a strawman commit that does that? It sounds like something unsound that needs a fix elsewhere.

Reply:
I can recreate on Monday. Really, I don't think that making the slice mut or not makes any difference. If I return a Rust object from the rest equivalent of OperatorFieldGet*, then the user will have the ability to capture the object in a mutable variable and change as they want. If I return an immutable ref or an opt holding the immutable ref, then the underlying rust object has to live somewhere.

Really, this comes down to if we want to allow users to get the vectors out of the fields of an object and change the contents of those vectors, as only the vectors present a way to change the underlying data after creation.

Side note: Are we 100% certain that transmute will be portable? The docs seem to imply otherwise?

Reply:
As for the mutability leak you've found, you can't have a non-mut accessor return a mutable reference (or an owned object that behaves like a mutable reference, which is the case if it points into mutable C objects). Currently, `VectorOpt::Some` holds an immutable reference so the vector inside can't be mutated.

```rust
let mut inputs = op_mass.inputs()?;
if let libceed::prelude::VectorOpt::Some(v) = inputs[1].vector() {
    v.set_value(3.);
}
```
fails with
```
error[E0596]: cannot borrow `*v` as mutable, as it is behind a `&` reference
  --> src/vector.rs:96:5
   |
41 | if let libceed::prelude::VectorOpt::Some(v) = inputs[1].vector() {
   |                                          - help: consider changing this to be a mutable reference: `&mut Vector<'_>`
42 |     v.set_value(3.);
   |     ^ `v` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error: aborting due to previous error
```

We may need a way for a user to mutate vectors in this context, but I think we want to localize it and use `RefCell` or otherwise checked exclusion. I don't think `op.inputs()` should return something that can mutate vectors.

Here's a revision without `transmute`, following ["This is the proper no-copy, unsafe way of "transmuting"](https://doc.rust-lang.org/std/mem/fn.transmute.html): 70141ce359b58a9ebdcec77190151f921783050e

Reply:
Right, but that's with returning a VectorOpt. That's why I switched to a returning a VectorOpt. But we need to keep the Vector that the VectorOpt is referencing somewhere. Which is why I created the more complex OperatorField.

Reply:
I just realized that currently we bypass the borrow checker and silently mutate the vectors for our passive fields. We should probably fix that.

Reply:
We have to keep it somewhere *if* it's more than just the C pointer. I'm not sure I understand the motivation for the mirroring of related objects from C into Rust. Like why does `OperatorCore<'a>` contain a `&'a Ceed` instead of just a `PhantomData` to hold the lifetime?

Mutability in passive outputs is a common pattern that I don't know an idiomatic solution to, other than runtime checking similar to the read/write locking checking in vectors. With smart references, this could look like reference counting in which there can only ever be one mutable container, but an arbitrary number of immutable containers that are capable of holding it until needed. For example, a setup Operator could create an output that is fed into other Operators (solvers and diagnostics). The setup Operator might be kept around for multiple iterations, say while mesh geometry is optimized, then dropped with its output still used (immutably). Way out of scope for this PR, but the exclusivity of the mutable container could be checked at compile time. (This might exist in crate form.)

Reply:
I don't understand how we would prevent mutation without hiding the ownership of the underlying C object from the user. If a field returns a C pointer to a vector or other object, the user can make that return value mut unless we borrow it from somewhere, unless I'm missing something.u

We carry a Rust Ceed in all objects to simplify error handling. Otherwise we would need to retrieve the C Ceed to check for errors after each bind_ceed call.

Reply:
They can make what `mut`? We either return an object that derefs to `Vector` (could just be wrapping the C pointer) or a reference to `Vector`. It's true that if we return an (owned) `Vector` then the user can do anything with it, but we shouldn't do that.

I'm not sure the error-checking is that different from having a method to access it. Saving a few instructions in an error handling path doesn't seem important. A possible benefit would be possibly better resilience to memory corruption, in which calling into C again could give an erroneous `Ceed`. That seems like an edge case.

As for lifetimes, you can use `PhantomData` to match lifetimes without holding a direct reference. But I don't even think that matters because the C `Ceed` is reference counted so it doesn't matter if the Rust `Ceed` is dropped while objects live on.

Reply:
If we return a reference, I don't understand where ownership lives. I'll just play around on Monday, because I'm more confused at this point and need to just tinker with code until I see what you're seeing.

Reply:
Ownership would live on the C side with reference counting. We can return a reference or slice or struct that implements `Deref` (all of which have a lifetime) without incrementing the reference count of the C object because Rust checks that these ref-like things cannot outlive the parent object (which holds a pointer to the C object with in-built reference counting). I think if you can point to a strawman of what you think is problematic, I'll be able to help.